src/libgap-api.h

```diff
@@ -3,9 +3,93 @@
 #ifndef LIBGAP_API_H
 #define LIBGAP_API_H
 
+#include "gasman.h"
 #include "system.h"
 
 
+#ifdef __GNUC__
+#define unlikely(x)     __builtin_expect(!!(x), 0)
+#else
+#define unlikely(x)     (x)
+#endif
+
+
+#ifndef GAP_ENTER_DEBUG
+#define GAP_ENTER_DEBUG 0
+#endif
+
+
+extern syJmp_buf * _GAP_GetReadJmpError(void);
+extern Int  _GAP_GetEnterStackCount(void);
+extern void _GAP_IncEnterStackCount(void);
+extern void _GAP_DecEnterStackCount(void);
+extern void _GAP_SetEnterStackCount(Int count);
+
+
+#if GAP_ENTER_DEBUG
+#define GAP_ENTER_DEBUG_MESSAGE(message, file, line) \
+    fprintf(stderr, "%s: %d; %s:%d\n", message, _GAP_EnterStackCount, file, line);
+#else
+#define GAP_ENTER_DEBUG_MESSAGE(message, file, line)
+#endif
+
+
+#define GAP_EnterStack() \
+    GAP_ENTER_DEBUG_MESSAGE("EnterStack", __FILE__, __LINE__); \
+    Int _gap_tmp_enter_stack_count = _GAP_GetEnterStackCount(); \
+    if (_gap_tmp_enter_stack_count < 0) { \
+        _GAP_SetEnterStackCount(-_gap_tmp_enter_stack_count); \
+    } else { \
+        if (_gap_tmp_enter_stack_count == 0) { \
+            MarkStackBottomBags(); \
+        } \
+        _GAP_IncEnterStackCount(); \
+    }
+
+
+#define GAP_LeaveStack() \
+    _GAP_DecEnterStackCount(); \
+    GAP_ENTER_DEBUG_MESSAGE("LeaveStack", __FILE__, __LINE__);
+
+
+static inline int _GAP_Error_Prejmp(const char* file, int line) {
+#if GAP_ENTER_DEBUG
+    GAP_ENTER_DEBUG_MESSAGE("Error_Prejmp", file, line);
+#endif
+    if (_GAP_GetEnterStackCount() > 0) {
+        return 1;
+    }
+    return 0;
+}
+
+
+static inline int _GAP_Error_Postjmp(int JumpRet)
+{
+    if (unlikely(JumpRet != 0)) {
```

Comment:
In this case the notion was also borrowed from cysignals, on which significantly more profiling has been done, and it can make a difference in some cases.  I think it's especially important since any code which uses GAP is going to have to be wrapped with these macros so might as well squeeze what we can out of them.

Reply:
But now that code is inside `src/libgap-api.c`, and so an expensive function call needs to be performed to get to it, so I somewhat doubt the benefit will manifest. But I'll add more comments on the latest diff, instead of this "outdated" one.

Reply:
You're right, actually, the `unlikely(...)` should go around the function call itself, not in the function (I think in some earlier version I had this was just a macro).